server/src/main/java/com/ncc/neon/services/ClusterService.java

```diff
@@ -23,16 +26,190 @@
+
+        // take care of extra keys that may be present
+        Map<String, Object> extraKeySets = getExtraKeySetsMap(fieldNameKey, aggregationNameKey, data);
+
+        // aggregate data into the new clusters
+        aggregateNumbersInNewData(fieldNameKey, aggregationNameKey, data, newData, extraKeySets);
+
+        this.clusterClause = null;
+        return new TabularQueryResult(newData);
+    }
+
+    /**
+     * Calculates the spacing between the different bins in the new data set results.
+     *
+     * @param fieldNameKey the key that corresponds with the field name
+     * @param firstGroup the first group of data in the original result data
+     * @param lastGroup the last group of data in the original result data
+     * @param count how many bins there should be in the new data set
+     * @param clusters clusters specified by the cluster clause if available
+     * @return the new data set ready to be filled
+     */
+    private List<Map<String, Object>> getNewDataBins(String fieldNameKey, BigDecimal firstGroup, BigDecimal lastGroup,
+                                                     int count, List<List<Object>> clusters) {
+        List<Map<String, Object>> newData = new ArrayList<>();
+        if (clusters == null) {
+            BigDecimal gap = (lastGroup.subtract(firstGroup)).divide(new BigDecimal(count));
+            BigDecimal step = new BigDecimal(".0001");
+            BigDecimal currentBin = firstGroup;
+            for (int i = 0; i < count; i++) {
+                LinkedHashMap map = new LinkedHashMap<>();
+                ArrayList range = new ArrayList<>();
+                if (i != 0) {
+                    range.add(currentBin.add(step));
+                } else {
+                    range.add(currentBin);
+                }
+                range.add(currentBin = currentBin.add(gap));
+                map.put(fieldNameKey, range);
+                newData.add(map);
+            }
+        } else {
+            for (int i = 0; i < clusters.size(); i++) {
+                LinkedHashMap map = new LinkedHashMap<>();
+                ArrayList range = new ArrayList();
+                range.add(new BigDecimal(clusters.get(i).get(0).toString()));
+                range.add(new BigDecimal(clusters.get(i).get(1).toString()));
+                map.put(fieldNameKey, range);
+                newData.add(map);
+            }
+        }
+        return newData;
+    }
+
+    /**
+     * Setup any extra keys
+     *
+     * @param fieldNameKey fieldname key given by the cluster clause
+     * @param aggregationNameKey aggregation name key given by the cluster clause
+     * @param data original result data
+     * @return the extra key sets map for storing additional keys
+     */
+    private Map<String, Object> getExtraKeySetsMap(String fieldNameKey, String aggregationNameKey,
+                                                   List<Map<String, Object>> data) {
```

Comment:
This will probably need to be changed from `String fieldNameKey` to `List<String> fieldNameKey` in the lat/lon cluster ticket.  I'll let you decide on whether you want to do that now or later.

Reply:
Later, so I have the changes in one place.

Reply:
Cool.